### ISSUE TITLE
`root_at_midpoint` methods fails on certain `TreeNode` objects

### ISSUE THREAD

**Issue by johnchase** [biocore/scikit-bio#1077]
```python 
from skbio import TreeNode
tree = TreeNode.read('rep_set.tre', format='newick')
tree.root_at_midpoint()
```

Fails with:

```TypeError: unorderable types: TreeNode() > TreeNode()```

I apologize, I am not able to reproduce this error with a smaller tree object, as I don't entirely understand the issue. @wasade I think you may understand what is specifically happening here?

**Comment by wasade** [comment]
The following will recreate it. The issue arises if two paths have an identical distance which causes `max` to evaluate `item[1]` to resolve the "max". 

```python
10:55:49 (mcdonadt@8086):gg_13_5> from skbio import TreeNode

10:55:56 (mcdonadt@8086):gg_13_5> from io import StringIO

10:56:01 (mcdonadt@8086):gg_13_5> s = StringIO("((a:1,b:1)c:2,(d:3,e:4)f:5)root;")

10:56:33 (mcdonadt@8086):gg_13_5> t = TreeNode.read(s)

10:56:37 (mcdonadt@8086):gg_13_5> t.root_at_midpoint()

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-f586e253faaa> in <module>()
----> 1 t.root_at_midpoint()

/Users/mcdonadt/rs/scikit-bio/skbio/tree/_tree.py in root_at_midpoint(self)
    840         """
    841         tree = self.copy()
--> 842         max_dist, tips = tree.get_max_distance()
    843         half_max_dist = max_dist / 2.0
    844

/Users/mcdonadt/rs/scikit-bio/skbio/tree/_tree.py in get_max_distance(self)
   2377             # child is encountered
   2378             try:
-> 2379                 self._set_max_distance()
   2380             except TreeError:  #
   2381                 return self._get_max_distance_singledesc()

/Users/mcdonadt/rs/scikit-bio/skbio/tree/_tree.py in _set_max_distance(self)
   2321                 else:
   2322                     print(n)
-> 2323                     tip_info = [(max(c.MaxDistTips), c) for c in n.children]
   2324
   2325                     dists = [i[0][0] for i in tip_info]

/Users/mcdonadt/rs/scikit-bio/skbio/tree/_tree.py in <listcomp>(.0)
   2321                 else:
   2322                     print(n)
-> 2323                     tip_info = [(max(c.MaxDistTips), c) for c in n.children]
   2324
   2325                     dists = [i[0][0] for i in tip_info]

TypeError: unorderable types: TreeNode() > TreeNode()
```

To highlight the specific issue independent of `TreeNode`:

```python
10:56:47 (mcdonadt@8086):gg_13_5> max([(1.0, object()), (1.1, object())])
Out[6]: (1.1, <object at 0x1003b4320>)

10:59:08 (mcdonadt@8086):gg_13_5> max([(1.0, object()), (1.0, object())])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-d579afadc379> in <module>()
----> 1 max([(1.0, object()), (1.0, object())])

TypeError: unorderable types: object() > object()
```

**Comment by ebolyen** [comment]
Are there any common (hopefully published) ways to break the tie?

**Comment by gregcaporaso** [comment]
Ran into this one again today.

**Comment by gregcaporaso** [comment]
Ran into it again today, it basically happens with every real tree that is provides, so this method should be considered broken. I think we just need to choose one at random to break the tie.

**Comment by jairideout** [comment]
@wasade has a PR in to fix this: #1208 Just waiting on one minor change then it's ready for merge.

**Comment by gregcaporaso** [comment]
Oops, didn't see that.

**Comment by gregcaporaso** [comment]
This does work (in Py3), just for the record:

```python
from ete3 import Tree
t = Tree("combined-ssu-align/combined-ssu-align.bacteria.mask.tre", format=1)
t.set_outgroup(t.get_midpoint_outgroup())
t.write(outfile="combined-ssu-align/combined-ssu-align.bacteria.mask.mid-rooted.tre", format=1)
```

**Comment by wasade** [comment]
Awesome! I'll be honest, I always had trouble with the cogent
implementation too, which the skbio one is of course derived from. The
traceback on the PR would still be useful but my _guess_ is recursion depth
limit

**Comment by gregcaporaso** [comment]
Yep:

```python
from skbio import TreeNode
t = TreeNode.read("combined-ssu-align/combined-ssu-align.bacteria.mask.tre")
t.root_at_midpoint()

...

/home/gregcaporaso/code/scikit-bio/skbio/tree/_tree.py in unrooted_copy(self, parent)
    600         """
    601         neighbors = self.neighbors(ignore=parent)
--> 602         children = [c.unrooted_copy(parent=self) for c in neighbors]
    603
    604         # we might be walking UP the tree, so:

/home/gregcaporaso/code/scikit-bio/skbio/tree/_tree.py in <listcomp>(.0)
    600         """
    601         neighbors = self.neighbors(ignore=parent)
--> 602         children = [c.unrooted_copy(parent=self) for c in neighbors]
    603
    604         # we might be walking UP the tree, so:

/home/gregcaporaso/code/scikit-bio/skbio/tree/_tree.py in unrooted_copy(self, parent)
    617
    618         result = self.__class__(name=edgename, children=children,
--> 619                                 length=length)
    620
    621         if parent is None:

RecursionError: maximum recursion depth exceeded
```

Is this something that could be fixed? If not, could we detect it and error (e.g., if there a max number of nodes in the tree that we can support, it would be great to just error and tell the user they are using a tree that's too big to support this operation with ``skbio.TreeNode``)?

**Comment by wasade** [comment]
Unrooted copy needs to be iterative. Iirc ete2 ups the recursion limit
which is not safe in the event of an errant process exploding the stack
(segfaults such to debug).

Could try/except, double recursion limit, try again then fail gracefully.
Iterative alg probably isn't hard but will take thought

**Comment by gregcaporaso** [comment]
I don't see a call to [``setrecursionlimit`` in the code base](https://github.com/jhcepas/ete/search?utf8=%E2%9C%93&q=setrecursionlimit&type=Code``) - why do you think it's doing that? Is there another way to increase the recursion limit?

**Comment by gregcaporaso** [comment]
I think we probably shouldn't worry about fixing it, but just detecting it and giving a better error message. We could just catch the ``RecursionError`` and point to a new issue about this that notes the workaround.

**Comment by wasade** [comment]
thought i saw it when digging through a 1.5yrs ago, either misremembering
or code revision

**Comment by wasade** [comment]
interesting, so ete isn't doing a deepcopy but modifying in place. if you
trigger a deepcopy with it, it should hit recursion.

**Comment by gregcaporaso** [comment]
@wasade, will you be able to address the comments on this PR? We'd need them by tomorrow to get into 0.4.1.

**Comment by gregcaporaso** [comment]
Oops, it looks like comments on this are mixed between this issue and the corresponding PR, #1208.
